src/unix/udp.c

```diff
@@ -148,6 +166,51 @@ static void uv__udp_io(uv_loop_t* loop, uv__io_t* w, unsigned int revents) {
   }
 }
 
+static void uv__udp_recvmmsg(uv_udp_t* handle, uv_buf_t* buf) {
+  struct sockaddr_in6 peers[MMSG_MAXWIDTH];
+  struct iovec iov[MMSG_MAXWIDTH];
+  struct uv__mmsghdr msgs[MMSG_MAXWIDTH];
+  ssize_t nread;
+  uv_buf_t chunk_buf;
+  int chunks;
+  int flags;
+  int k;
+
+  /* prepare structures for recvmmsg */
+  chunks = buf->len / UDP_DGRAM_MAXSIZE;
+  if (chunks > MMSG_MAXWIDTH)
+    chunks = MMSG_MAXWIDTH;
+  for (k = 0; k < chunks; ++k) {
+    iov[k].iov_base = buf->base + k * UDP_DGRAM_MAXSIZE;
+    iov[k].iov_len = UDP_DGRAM_MAXSIZE;
+    msgs[k].msg_hdr.msg_iov = iov + k;
+    msgs[k].msg_hdr.msg_iovlen = 1;
+    msgs[k].msg_hdr.msg_name = peers + k;
+    msgs[k].msg_hdr.msg_namelen = sizeof(peers[0]);
+  }
+
+  do {
+    nread = uv__recvmmsg(handle->io_watcher.fd, msgs, chunks, 0, NULL);
+  } while (nread == -1 && errno == EINTR);
+
+  if (nread < 1) {
+    if (nread == 0 || errno == EAGAIN || errno == EWOULDBLOCK)
+      handle->recv_cb(handle, 0, buf, NULL, 0);
+    else
+      handle->recv_cb(handle, UV__ERR(errno), buf, NULL, 0);
+  }
+  else {
+    /* count to zero, so the buffer base comes last */
+    for (k = nread; k-- && handle->recv_cb != NULL;) {
+      flags = 0;
+      if (msgs[k].msg_hdr.msg_flags & MSG_TRUNC)
+        flags |= UV_UDP_PARTIAL;
+
+      chunk_buf = uv_buf_init(iov[k].iov_base, iov[k].iov_len);
+      handle->recv_cb(handle, msgs[k].msg_len, &chunk_buf, msgs[k].msg_hdr.msg_name, flags);
```

Comment:
Generally alloc_cb and recv_cb come in pairs, wouldn't this break that assumption?

Reply:
It does. I added a special flag passed to recv_cb to signal that this buffer is a part of a larger buffer and must not be freed.

Reply:
👍